### DISCUSSION TITLE
How to center a Window?

### DISCUSSION THREAD

**Discussion by abolzoniSpringa** [AvaloniaUI/Avalonia#8048] [Q&A] [answered]
Hi,

I'm trying to add some popups to my application, but every time I open one  it's in a differente position of the screen.
The position changes slightly but is always in the top left area of the screen.
Setting WindowStartupLocation as CenterScreen or CenterOwner has no effect.
Reading gitter I found out that there may be some issues on MacOs or Linux, but I'm not able to center it even on windows.

The following is an example of a small popup window, opened using Show(this) from a maximized parent window.
Is there a better way to center a window?

`<Window xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        mc:Ignorable="d" d:DesignWidth="320" d:DesignHeight="160"
        x:Class="SlingshotAvalonia.Views.Popup"
		xmlns:local="clr-namespace:SlingshotAvalonia"
		CanResize="False"
		Background="{x:Null}"
		ExtendClientAreaToDecorationsHint="True"
		ExtendClientAreaTitleBarHeightHint="999"
		ExtendClientAreaChromeHints="NoChrome"
		TransparencyLevelHint="AcrylicBlur"
		Width="320" Height="160"
		Title="Popup" FontFamily="Source Sans Pro" FontSize="14"  Topmost="True" WindowStartupLocation="CenterScreen">`

**Comment by MasterMann** [comment]
Unfortunately, it seems like the release version has a bug. Tested it on Linux - window appears in top-right/bottom-right randomly.
I made it work manually (by centering the window after it activates, computing the needed offsets from left and top of the screen).

**Comment by timunie** [comment]
On Windows I can't reproduce this issue. Do you have a small sample for this?

**Reply by abolzoniSpringa** [reply]
My fault, on the sample the center command is working, only on my application this happens.
I'll try to investigate more on what may be causing this.
Thanks

**Reply by abolzoniSpringa** [reply]
Actually I was wrong.

Show(this) works as inteded.

It's showDialog(ownerWindow) that fails to center.
It always shows up in the top left area of the screen.

as for a sample I made a simple empty application with only a button. 
Clicking it triggers the following:

  if (Application.Current.ApplicationLifetime is IClassicDesktopStyleApplicationLifetime desktop)
                {
                    new Window1().ShowDialog(desktop.MainWindow);
                }

window1 is a small empty window, which shows up offcenter.

**Reply by timunie** [reply]
Also this works for me. A tiny sample App on GitHub may help. Keep in mind, I tested with Windows OS only. Mac and Linux may behave differently. 

![image](https://user-images.githubusercontent.com/47110241/166683912-4b14cf6c-8b10-4df1-98b1-659902cbecc4.png)

**Reply by abolzoniSpringa** [reply]
Ok, I finally found the issue,

My window is set to WindowState="Maximized"
this causes every popup to be offset.
If i delete it, it works fine, but my window is not filling the screen.

This is the same if I click the restore button on the window with a default windowsState.
as soon as the window changes size, the popup are no more centered.

**Reply by timunie** [reply]
The popup window will not automatically follow your window. So once the window is open, the position stays the same. Unless you do it manually. 

Again, a simple repo can often help. Maybe a screen recoding can help as well.

**Reply by abolzoniSpringa** [reply]
thank you, 
I'm ok with the popup not following the window, but I thougth that by setting it to maximized there'd be no need to follow the window since it is already the full size.

I made a small repo with a mainwindow and a popup window
https://github.com/abolzoniSpringa/AvaloniaTestCenterDialog

Here I will try to center the popup on opening manually.

**Comment by timunie** [comment, marked as answer]
Seems like you were hitting a bug in 0.10.12. If I update to 0.10.14 it works well. https://github.com/abolzoniSpringa/AvaloniaTestCenterDialog/pull/1 

Happy coding 
Tim

**Reply by abolzoniSpringa** [reply]
thanks again for your help! now it works!

**Comment by suwet** [comment]
Calculate position when Onpened event of window.

in constructor of your window class 
 this.Opened += OnOpened;

 private void OnOpened(object? sender, EventArgs e) {
            int window_w = (int)this.DesiredSize.Width / 2;
            int window_h = (int)this.DesiredSize.Height / 2;
            int x = (int)(Program.MainWindow.Bounds.Width / 2) - window_w;
            int y = (int)(Program.MainWindow.Bounds.Height / 2) - window_h;
            this.Position = new Avalonia.PixelPoint(x, y);
        }

It work for me.
